Sources/X509/Verifier/VerificationDiagnostic.swift

```diff
@@ -0,0 +1,237 @@
+        _ leafCertificate: Certificate,
+        reason failsToMeetPolicyReason: String
+    ) -> Self {
+        .leafCertificateIsInTheRootStoreButDoesNotMeetPolicy(.init(
+            leafCertificate: leafCertificate,
+            failsToMeetPolicyReason: failsToMeetPolicyReason
+        ))
+    }
+    static func chainFailsToMeetPolicy(
+        _ chain: UnverifiedCertificateChain,
+        reason failsToMeetPolicyReason: String
+    ) -> Self {
+        .chainFailsToMeetPolicy(.init(
+            chain: chain,
+            failsToMeetPolicyReason: failsToMeetPolicyReason
+        ))
+    }
+    static func issuerHashUnhandledCriticalExtension(
+        issuer: Certificate,
+        partialChain: [Certificate],
+        handledCriticalExtensions: [ASN1ObjectIdentifier]
+    ) -> Self {
+        .intermediateHashUnhandledCriticalExtension(.init(
+            issuer: issuer,
+            partialChain: partialChain,
+            handledCriticalExtensions: handledCriticalExtensions
+        ))
+    }
+    static func issuerHasNotSignedCertificate(
+        _ issuer: Certificate,
+        partialChain: [Certificate]
+    ) -> Self {
+        .issuerHasNotSignedCertificate(.init(
+            issuer: issuer,
+            partialChain: partialChain
+        ))
+    }
+}
+
+extension VerificationDiagnostic: CustomStringConvertible {
+    /// Produces a human readable description of this ``VerificationDiagnostic`` that is potential expensive to compute.
+    public var description: String {
+        storage.description
+    }
+}
+
+extension VerificationDiagnostic.Storage {
+    var description: String {
+        switch self {
+        case .leafCertificateHasUnhandledCriticalExtension(let diagnostic): return diagnostic.description
+        case .leafCertificateIsInTheRootStoreButDoesNotMeetPolicy(let diagnostic): return diagnostic.description
+        case .chainFailsToMeetPolicy(let diagnostic): return diagnostic.description
+        case .intermediateHashUnhandledCriticalExtension(let diagnostic): return diagnostic.description
+        case .issuerHasNotSignedCertificate(let diagnostic): return diagnostic.description
+        }
+    }
+}
+
+extension VerificationDiagnostic.LeafCertificateHasUnhandledCriticalExtensions: CustomStringConvertible {
+    var description: String {
+        """
+        The leaf certificate has critical extensions that the policy does not understand and therefore can't enforce.
+        Unhandled extensions:
+        \(leafCertificate.extensions.unhandledCriticalExtensions(
+            for: self.handledCriticalExtensions
+        ).lazy.map { $0.description }.joined(separator: "\n"))
+        Leaf certificate:
+        \(leafCertificate)
+        """
+    }
+}
+
+extension VerificationDiagnostic.LeafCertificateIsInTheRootStoreButDoesNotMeetPolicy: CustomStringConvertible {
+    var description: String {
+        """
+        Leaf certificate is in the root store of the verifier but it does by itself not meet the policy.
```

Comment:
```suggestion
        Leaf certificate is in the root store of the verifier but it does not meet the policy.
```

Reply:
I kinda want to highlight that the leaf certificate may only not meet the policy if the whole chain is just this single certificate but may pass the policy if we try a different chain e.g. through intermediates. We don't know that yet though, just that it is in the root store and the policy has failed on that single certificate chain. We will continue trying to find new paths through intermediates to a root which can then still pass the policy.